modules/layers/src/screen-grid-layer/screen-grid-layer.js

```diff
@@ -187,15 +187,13 @@ export default class ScreenGridLayer extends Layer {
   getPickingInfo({info, mode}) {
     const {index} = info;
     if (index >= 0) {
-      let {aggregationResults} = this.state;
-      if (!aggregationResults) {
-        aggregationResults = {
-          aggregationData: this.state.aggregationBuffer.getData(),
-          maxData: this.state.maxBuffer.getData()
-        };
-        // Cache aggregationResults to avoid multiple buffer reads.
-        this.setState({aggregationResults});
-      }
+      const {aggregationResults} = this.state;
+
+      // Cache aggregationResults to avoid multiple buffer reads.
+      aggregationResults.aggregationData =
+        aggregationResults.aggregationData || this.state.aggregationBuffer.getData();
```

Comment:
Shouldn't the array vs. buffer difference be handled inside `GPUAggregator`? Do all users of the class need to be aware of this caveat?

Reply:
Good point.
Couple of things going on here:

1. GPUAggregator always returns `Buffer` objects, whether it is WebGL1 or WebGL2, but if the user wants to read the results into CPU memory, it needs to know whether it is under WebGL1 or WebGL2 because, it under WebGL1, buffer.getData() is not supported, and also under WebGL1, data is aggregated by CPU and results are available in JS Arrays. So here I am checking if the results already contains JS Arrays (WebGL1) , if not I am reading it from `Buffer` object (WebGL2).
2. Once aggregation is completed, and if the application is using WebGL2, we only need to read data from buffer once and cache it (invalidated when data is re-aggregated).

I will address this in a separate PR (#2561).